### PR TITLE
add TRACE logging via -dd

### PR DESCRIPTION
**Pull request by mr-tz**
addresses #413

This is a start, I expect we will tweak this moving forward.

### CODE REVIEW

File: floss/api_hooks.py [hunk 1]
```diff
@@ -17,19 +21,20 @@ def __init__(self, vw, function_index):
 
     def apicall(self, emu, op, pc, api, argv):
         # overridden from Monitor
-        self.d("apicall: %s %s %s %s %s", emu, op, pc, api, argv)
```
Comment (mr-tz) [inline floss/api_hooks.py:20]: `emu` doesn't provide any value here

File: floss/api_hooks.py [hunk 2]
```diff
@@ -17,19 +21,20 @@ def __init__(self, vw, function_index):
...
+        logger.trace("0x%x %s %s %s", pc, op, api, argv)
 
     def prehook(self, emu, op, startpc):
         # overridden from Monitor
-        pass
+        # helpful for debugging decoders, but super verbose!
+        logger.trace("0x%x %s", startpc, op)
```
Comment (mr-tz) [inline floss/api_hooks.py:29]: may want to disable this even at this level?!

File: floss/api_hooks.py [hunk 3]
```diff
@@ -239,6 +245,12 @@ def hook(self, callname, driver, callconv, api, argv):
             va = self._allocate_mem(driver, size)
             callconv.execCallReturn(driver, va, len(argv))
             return True
+        elif callname == "_calloc_base":
+            size = argv[0]
+            count = argv[1]
+            va = self._allocate_mem(driver, size * count)
+            callconv.execCallReturn(driver, va, 2)  # TODO len(argv)?
+            return True
```
Comment (mr-tz) [inline floss/api_hooks.py:253]: sorry, unrelated addition :/

File: floss/identify.py [hunk 4]
```diff
@@ -112,53 +111,54 @@ def find_decoding_function_features(vw, functions, disable_progress=False):
...
-        function_data["features"].append(BlockCount(function_data["meta"].get("block_count")))
-        function_data["features"].append(InstructionCount(function_data["meta"].get("instruction_count")))
-        function_data["features"].append(Arguments(function_data["meta"].get("api", []).get("arguments")))
-
-        for feature in extract_function_features(f):
-            function_data["features"].append(feature)
-
-        for bb in f.basic_blocks:
-            for feature in extract_basic_block_features(f, bb):
+    with tqdm.contrib.logging.logging_redirect_tqdm(), floss.utils.redirecting_print_to_tqdm():
```
Comment (mr-tz) [inline floss/identify.py:114]: allows to log and print nicely while using `tqdm`

File: floss/logging.py [hunk 5]
```diff
@@ -1,7 +1,12 @@
 import logging
 
+TRACE = logging.DEBUG - 1
+setattr(logging, "TRACE", TRACE)
+logging.addLevelName(TRACE, "TRACE")
+
 GREY = "\x1b[38;21m"
```
Comment (mr-tz) [inline floss/logging.py:7]: maybe use `colorama` instead?

File: floss/main.py [hunk 6]
```diff
@@ -307,12 +284,39 @@ def make_parser(argv):
...
 
+    output_group = parser.add_argument_group("rendering arguments")
+    output_group.add_argument("-j", "--json", action="store_true", help="emit JSON instead of text")
+    output_group.add_argument(
+        "-v", "--verbose", action="store_true", help="enable verbose result document (no effect with --json)"
+    )
+    output_group.add_argument(
+        "-vv", "--vverbose", action="store_true", help="enable very verbose result document (no effect with --json)"
+    )
+    # TODO currently unused
```
Comment (mr-tz) [inline floss/main.py:295]: see #462

File: floss/string_decoder.py [hunk 7]
```diff
@@ -219,23 +219,7 @@ def extract_strings(b: DeltaBytes, min_length, no_filter) -> List[DecodedString]
...
-            decoded_string = floss.utils.strip_string(s.string)
-        else:
-            continue
-
-        if len(decoded_string) >= min_length:
-            ret.append(
-                DecodedString(b.address + s.offset, b.address_type, decoded_string, b.decoded_at, b.decoding_routine)
-            )
-
-    for s in floss.strings.extract_unicode_strings(b.bytes):
```
Comment (mr-tz) [inline floss/string_decoder.py:238]: This was all 1:1 the same code as above...

File: floss/logging.py [hunk 8]
```diff
@@ -13,6 +18,7 @@ def make_format(color):
 
 
 FORMATS = {
+    logging.TRACE: make_format(MAUVE),  # type: ignore
```
Comment (mr-tz) [inline floss/logging.py:21]: is this ok to make `mypy` happy?

File: floss/main.py [hunk 9]
```diff
@@ -373,7 +377,8 @@ def select_functions(vw, asked_functions: Optional[List[int]]) -> Set[int]:
     if missing_functions:
         raise ValueError("failed to find functions: %s" % (", ".join(map(hex, sorted(missing_functions)))))
 
-    logger.debug("selected the following functions: %s", ", ".join(map(hex, sorted(asked_functions_))))
+    logger.debug(f"selected %d functions", len(asked_functions_))
```
Comment (williballenthin) [inline floss/main.py:380]:
```suggestion
    logger.debug("selected %d functions", len(asked_functions_))
```

File: floss/main.py [hunk 10]
```diff
@@ -513,9 +518,9 @@ def print_stack_strings(extracted_strings: Union[List[StackString], List[TightSt
 
 
 def print_file_meta_info(vw, selected_functions: Set[int]):
-    logger.info("analysis summary:")
+    logger.trace("analysis summary:")
     for k, v in get_vivisect_meta_info(vw, selected_functions).items():
-        logger.info("  %s: %s" % (k, v or "N/A"))
+        logger.trace("  %s: %s" % (k, v or "N/A"))
```
Comment (williballenthin) [inline floss/main.py:523]:
```suggestion
        logger.trace("  %s: %s",  k, v or "N/A")
```

File: floss/main.py [hunk 11]
```diff
@@ -715,9 +720,9 @@ def main(argv=None) -> int:
             logger.error(e.args[0])
             return -1
 
-        logger.info("analysis summary:")
+        logger.trace("analysis summary:")
         for k, v in get_vivisect_meta_info(vw, selected_functions).items():
-            logger.info("  %s: %s" % (k, v or "N/A"))
+            logger.trace("  %s: %s" % (k, v or "N/A"))
```
Comment (williballenthin) [inline floss/main.py:725]:
```suggestion
            logger.trace("  %s: %s", k, v or "N/A")
```

**Review by williballenthin — APPROVED** [review]
nice!

**Pull request merged**